File: app/src/main/java/com/rikkei/tranning/chatapp/views/uis/signup/SignUpViewModel.java
package com.rikkei.tranning.chatapp.views.uis.signup;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.rikkei.tranning.chatapp.services.models.UserModel;
import com.rikkei.tranning.chatapp.services.models.LoginUserModel;
import com.rikkei.tranning.chatapp.services.repositories.SignUpRepositories;

public class SignUpViewModel extends ViewModel {
    FirebaseAuth firebaseAuth= FirebaseAuth.getInstance();
    public MutableLiveData<String> userEmail = new MutableLiveData<>();
    public MutableLiveData<String> userPass = new MutableLiveData<>();
    public MutableLiveData<String> userName=new MutableLiveData<>();
    public MutableLiveData<Boolean> signUpSuccess=new MutableLiveData<>();
    public MutableLiveData<LoginUserModel> signUpUserMutableLiveData=new MutableLiveData<>();
    public void signUpButtonOnClick(){
        LoginUserModel user=new LoginUserModel(userEmail.getValue(),userPass.getValue());
        signUpUserMutableLiveData.setValue(user);
    }
//    public void replaceFragmentClick(){
//        getNavigator().replaceFragment();
//    }
//    public boolean validateEmailPassword(String email, String password){
//        if(Pattern.compile(
//                "^(([\\w-]+\\.)+[\\w-]+|([a-zA-Z]|[\\w-]{2,}))@"
//                        + "((([0-1]?[0-9]{1,2}|25[0-5]|2[0-4][0-9])\\.([0-1]?"
//                        + "[0-9]{1,2}|25[0-5]|2[0-4][0-9])\\."
//                        + "([0-1]?[0-9]{1,2}|25[0-5]|2[0-4][0-9])\\.([0-1]?"
//                        + "[0-9]{1,2}|25[0-5]|2[0-4][0-9]))|"
//                        + "([a-zA-Z]+[\\w-]+\\.)+[a-zA-Z]{2,4})$"
//        ).matcher(email).matches() && Pattern.compile("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)[a-zA-Z\\d]{8,}$"
//        ).matcher(password).matches()){
//            return true;
//        }
//        else {
//            return false;
//        }
//    }
//    public void checkboxClick(){
//        getNavigator().setEnableButton();
//    }
//    public TextWatcher nameTextWatcher(){
//        return new TextWatcher() {
//            @Override
//            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//            }
//            @Override
//            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//                getNavigator().setEnableButton();
//            }
//            @Override
//            public void afterTextChanged(Editable editable) {
//                if(TextUtils.isEmpty(editable)){
//                    getNavigator().requireName("Error");
//                }
//            }
//        };
//    }
//    public TextWatcher emailTextWatcher(){
//        return new TextWatcher() {
//            @Override
//            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//            }
//            @Override
//            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//                getNavigator().setEnableButton();
//            }
//            @Override
//            public void afterTextChanged(Editable editable) {
//            }
//        };
//    }
//    public TextWatcher passTextWatcher(){
//        return  new TextWatcher() {
//            @Override
//            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//            }
//            @Override
//            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//                getNavigator().setEnableButton();
//            }
//            @Override
//            public void afterTextChanged(Editable editable) {
//            }
//        };
//    }
//    public void signUpClick(){
//        getNavigator().signup();
//    }
    public void createUserFireBase(){
        firebaseAuth.createUserWithEmailAndPassword(userEmail.getValue(),userPass.getValue())
                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                    @Override
                    public void onSuccess(AuthResult authResult) {
                        FirebaseUser firebaseUser=firebaseAuth.getCurrentUser();
                        String userId=firebaseUser.getUid();
                        UserModel user=new UserModel(userId,userName.getValue(),userEmail.getValue(),"default","default","default");
                        new SignUpRepositories().createUserFromFirebase(user,userId);
                        //getNavigator().showMessageSignUp("Sign Up Success!");
                        signUpSuccess.setValue(true);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        signUpSuccess.setValue(false);
                    }
                });
    }
}

File: app/src/main/java/com/rikkei/tranning/chatapp/views/uis/signup/SignUpViewModel.kt
package com.rikkei.tranning.chatapp.views.uis.signup

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.rikkei.tranning.chatapp.services.models.LoginUserModel
import com.rikkei.tranning.chatapp.services.models.UserModel
import com.rikkei.tranning.chatapp.services.repositories.SignUpRepository

class SignUpViewModel : ViewModel() {
    var firebaseAuth = FirebaseAuth.getInstance()
    var userEmail = MutableLiveData<String>()
    var userPass = MutableLiveData<String>()
    var userName = MutableLiveData<String>()
    var signUpSuccess = MutableLiveData<Boolean>()
    var signUpUserMutableLiveData = MutableLiveData<LoginUserModel>()

    var signUpStatus = MutableLiveData<SignUpStatus>()

    fun signUpButtonOnClick() {
        val user = LoginUserModel(userEmail.value!!, userPass.value!!)
//        signUpStatus.value = SignUpStatus.Loading(true)
//        val user = LoginUserModel(userEmail.value.toString(), userPass.value.toString())
        if (user.isEmailValid)
            signUpUserMutableLiveData.value = user

        if (!user.validateEmailPassword()) {
            signUpStatus.value = SignUpStatus.errorPassAndEmail(true)
        } else {
            createUserFireBase()
        }

    }

    fun createUserFireBase() {

        signUpStatus.value = SignUpStatus.Loading(true)
        firebaseAuth.createUserWithEmailAndPassword(
            userEmail.value!!,
            userPass.value!!
        )
            .addOnSuccessListener {
                val firebaseUser = firebaseAuth.currentUser
                val userId = firebaseUser!!.uid
                val user = UserModel(
                    userId,
                    userName.value,
                    userEmail.value,
                    "default",
                    "default",
                    "default"
                )
                SignUpRepository()
                    .createUserFromFirebase(user, userId)
                //getNavigator().showMessageSignUp("Sign Up Success!");
                signUpSuccess.value = true
                signUpStatus.value = SignUpStatus.isOk(true)
                signUpStatus.value = SignUpStatus.Loading(false)


            }
            .addOnFailureListener {
                signUpStatus.value = SignUpStatus.Failure(it)
                signUpStatus.value = SignUpStatus.isOk(false)
                signUpStatus.value = SignUpStatus.Loading(false)

            }
    }

    sealed class SignUpStatus {

        data class Loading(var loading: Boolean) : SignUpStatus()

        data class isOk(var isLogin: Boolean) : SignUpStatus()

        data class Failure(var e: Throwable) : SignUpStatus()

        data class ErrorPassAndEmail(var isError: Boolean) : SignUpStatus()

        companion object {

            fun loading(isLoading: Boolean): SignUpStatus = Loading(isLoading)

            fun success(isLogin: Boolean): SignUpStatus = isOk(isLogin)

            fun failure(e: Throwable): SignUpStatus = Failure(e)

            fun errorPassAndEmail(isErrorPassAndEmail: Boolean) =
                ErrorPassAndEmail(isErrorPassAndEmail)
        }

    }
}